include/hip/spirv_hip_fp16.h

```diff
@@ -1528,7 +1530,7 @@ THE SOFTWARE.
             __half hrcp(__half x)
             {
                 return __half_raw{
-                    __llvm_amdgcn_rcp_f16(static_cast<__half_raw>(x).data)};
+                    __llvm_rcp_f16(static_cast<__half_raw>(x).data)};
```

Comment:
```
  This file is an almost verbatim copy of
  hipamd /include/hip/amd_detail/amd_hip_fp16.h (revision 348a177).

  Let's try to keep it as close to upstream and possible. No clang-format,
  and minimal modifications.
```
This change steers away from being close to upstream.

Reply:
> minimal modifications

What do you suggest and why do you think we should have any references to anything AMD-specific?

Reply:
Leave it and also the defines above as they were. The defines make sure AMD builtins does not end up into device binary.

Reply:
it fails to build. Again, why should we have `amdgcn` references in a file called `spirv_`? Does not make any sense to me.

Reply:
> it fails to build.

Could you elaborate how the build fails? 


> why should we have amdgcn references in a file called spirv_?

Minimal changes makes it easier to apply changes from the upstream (when/if we need) by having less merge conflicts to be dealt with.

Reply:
> Could you elaborate how the build fails?

references to amdgcn instrinsics. You can look at the logs yourself on github actions. 

> Minimal changes makes it easier to apply changes from the upstream (when/if we need) by having less merge conflicts to be dealt with.

these are minimal changes. 

Still haven't heard a good reason for having references to amdgcn in a spirv file.

Reply:
> references to amdgcn instrinsics. You can look at the logs yourself on github actions.

```
2023-08-23T14:47:21.6360757Z /home/pvelesko/actions-runner/_work/chipStar/chipStar/include/hip/spirv_math_fwd.h:77:48: error: expected ')'
2023-08-23T14:47:21.6361234Z     __device__ __attribute__((const)) _Float16 __llvm_amdgcn_rcp_f16(_Float16);
2023-08-23T14:47:21.6361597Z                                                ^
2023-08-23T14:47:21.6362326Z /home/pvelesko/actions-runner/_work/chipStar/chipStar/include/hip/spirv_hip_fp16.h:30:45: note: expanded from macro '__llvm_amdgcn_rcp_f16'
2023-08-23T14:47:21.6362719Z #define __llvm_amdgcn_rcp_f16(__X) ((__half)1.0f / (__half)__X)
...
```
The compiler error isn't about use of amdgcn intrinsic but a syntax error. And that happen because the `__llvm_amdgcn_rcp_f16` gets expanded by the macro from spirv_hip_fp16.h which is meant to replace calls and it breaks on function declarations. Header inclusion order change could explain why the error appears now.  The inclusion order is not the problem but the __llvm_amdgcn_rcp* defines in the beginning of spirv_hip_fp16.h header are not meant to be defined outside file. You can fix it by undefining them at the end of the spirv_hip_fp16.h:
```
#undef __llvm_amdgcn_rcp_f16
#undef __llvm_amdgcn_rcp_2f16
```

> Still haven't heard a good reason for having references to amdgcn in a spirv file.

I'm not in favor having amdgcn references in chipStar's HIP headers. The intent of having minimal changes on this file is to making applying changes from upstream with less trouble.

Reply:
> You can fix it by undefining them at the end of the spirv_hip_fp16.h:

or by removing all references to "amdgcn" - am I missing something here?

> I'm not in favor having amdgcn references in chipStar's HIP headers

I'd much rather not have any references to amdgcn than worry about changing a few lines next time we copy this file (not even clear if and when this would happen)

Since the beginning of this project, this header was copied once, maybe twice. I think so we're splitting hairs over what amounts to a couple of changed lines.

Reply:
Given the low amount of amdgcn references I'm in favor for removing them. CCing @pjaaskel to make sure he's fine with this too.

Reply:
Sure!